dace/transformation/passes/symbol_propagation.py

```diff
@@ -0,0 +1,214 @@
+        for cfgb, parent in all_cfgb.items():
+            self._update_syms(cfgb, parent, in_syms, out_syms)
+        return set()
+
+    # Given a CFGB, builds the incoming set of symbols
+    def _get_in_syms(
+        self,
+        sdfg: SDFG,
+        cfgb: ControlFlowBlock,
+        parent: ControlFlowRegion,
+        in_syms: Dict[ControlFlowBlock, Dict[str, Any]],
+        out_syms: Dict[ControlFlowBlock, Dict[str, Any]],
+    ) -> Dict[str, Any]:
+        # Combine the outgoing symbols of all incoming edges with their assignments to the CFGB
+        new_in_syms = {}
+        for i, edge in enumerate(parent.in_edges(cfgb)):
+            sym_table = copy.deepcopy(out_syms[edge.src])
+            sym_table.update(edge.data.assignments)
+
+            # Filter out symbols containing arrays accesses as they cannot be safely propagated (nested array accesses are not supported)
+            sym_table = {k: v for k, v in sym_table.items() if v is None or ("[" not in v and "]" not in v)}
+
+            # Also filter out symbols containing views as they cannot be safely propagated (they are seen as pointers)
+            sym_table = {
+                k: v
+                for k, v in sym_table.items() if v is None or not any([
+                    str(s) in sdfg.arrays and isinstance(sdfg.arrays[str(s)], dt.View)
+                    for s in pystr_to_symbolic(v).free_symbols
+                ])
+            }
+
+            # Combine the symbols
+            if i == 0:
+                new_in_syms = sym_table
+            else:
+                self._combine_syms(new_in_syms, sym_table)
+
+        # Nested starting CFBGs should inherit the symbols from their parent
+        # Ignore SDFGs as nested SDFGs have symbol mappings
+        if (parent.start_block == cfgb and not isinstance(parent, SDFG)) or (isinstance(parent, ConditionalBlock)
+                                                                             and cfgb in parent.sub_regions()):
+            assert new_in_syms == {}
+            new_in_syms = in_syms[parent]
+
+            # For LoopRegions, remove loop carried variables from the incoming symbols
+            if isinstance(parent, LoopRegion):
+                new_in_syms = copy.deepcopy(new_in_syms)
+                all_syms = set([s for e in parent.all_interstate_edges() for s in e.data.assignments.keys()])
+                for sym in all_syms:
+                    if sym in new_in_syms:
+                        new_in_syms[sym] = None
+
+        return new_in_syms
+
+    # Given a CFGB, builds the outgoing set of symbols
+    def _get_out_syms(
+        self,
+        cfgb: ControlFlowBlock,
+        parent: ControlFlowRegion,
+        in_syms: Dict[ControlFlowBlock, Dict[str, Any]],
+        out_syms: Dict[ControlFlowBlock, Dict[str, Any]],
+    ) -> Dict[str, Any]:
+        if isinstance(cfgb, LoopRegion):
+            # Any symbol that is assigned in the loop region is not propagated out
```

Comment:
Wait, you mean assigned through the update/init statements? Because symbols assigned on interstate edges definitely propagate out, we don't consider a nested control flow graph an isolated symbol scope.

Reply:
What is meant by this comment is that SymbolPropagation does not propagate the assignments inside loops to the outside. Imagine a loop with `a = some_val` in an interstate edge inside and `b = a` in an interstate edge after the loop. Performing the replacement `b = some_val` would be incorrect, since we don't know if the loop will be executed at all and thus, we don't know which value `a` has.